test/span.js

```diff
@@ -185,31 +184,80 @@ describe('span should', () => {
         assert.isOk(unnormalizedKey in Span._getBaggageHeaderCache());
     });
 
-    describe('with deferred sampling', () => {
-        beforeEach(function () {
-            spanContext._flags = constants.DEFERRED_SAMPLING_MASK;
-        });
-        it('should not pass deferred sampling flag to child spans', () => {
-            let child = tracer.startSpan('child', {childOf: span.context()});
-            assert.notEqual(child.context.flags & constants.DEFERRED_SAMPLING_MASK,
-                            constants.DEFERRED_SAMPLING_MASK);
-            assert.isOk(child._spanContext.samplingFinalized);
+    describe('with upsampling', () => {
+        beforeEach(() => {
+            tracer._upsampling = {enabled: true};
         });
 
-        it('should make a call to the underlying sampler and use the sampling decision when true', () => {
-            let mockSampler = sinon.mock(tracer._sampler);
-            mockSampler.expects('isSampled').withExactArgs('goodOperation', {}).returns(true);
-            let child = tracer.startSpan('goodOperation', {childOf: span.context()});
-            mockSampler.verify();
-            assert.isOk(child.context().isSampled());
-        });
+        function makeTestCase(flags, callSampler, originalSamplingDecision, originalOperation, newSamplingDecision, newOperation, finalSamplingDecision) {
+            return {
+                flags: flags,
+                callSampler: callSampler,
+                name: `Should ${callSampler?"":"not "}call sampler for flag ${flags} ${newOperation?"with modified operationName and samplingDecision":""}`,
+                original: {
+                    operationName: originalOperation,
+                    samplingDecision: originalSamplingDecision,
+                },
+                new: {
+                    operationName: newOperation,
+                    samplingDecision: newSamplingDecision,
+                },
+                finalSamplingDecision: finalSamplingDecision
+            }
+        }
+        const testCases = [
+            makeTestCase(0, true, true, "idli", null, null, true),
+            makeTestCase(0, true, false, "idli", null, null, false),
+            makeTestCase(0, true, true, "poriyal", true, "murungakkai", true),
+            // We ignore the following case because setOperationName can only go from a sampling decision of 0->1
+            // makeTestCase(0, true, true, "poriyal", false, "murungakkai", false),
```

Comment:
even more reason to have a test - we certainly want to test the scenario where downsampling "would have happened" and make sure it doesn't